from fastapi import Depends, HTTPException, status
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from jose import jwt, JWTError
from sqlalchemy.orm import Session
from internal.domain.person import Person
from internal.infrastructure.database.db import get_db

SECRET_KEY = "omsfdaioasdjioasiodsa24390234903209809u"
ALGORITHM = "HS256"

security = HTTPBearer()

def get_current_user(
    credentials: HTTPAuthorizationCredentials = Depends(security),
    db: Session = Depends(get_db)
):
    token = credentials.credentials
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        # âœ… Aceptamos tanto 'email' como 'sub' (por compatibilidad)
        email = payload.get("email") or payload.get("sub")

        if not isinstance(email, str):  # ðŸ‘ˆ verificamos tipo
            raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Token invÃ¡lido")

    except JWTError:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Token invÃ¡lido")

    user = db.query(Person).filter(Person.email == email).first()
    if not user:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Usuario no encontrado")

    return user
